firedrake/utility_meshes.py

```diff
@@ -659,6 +660,85 @@ def UnitIcosahedralSphereMesh(refinement_level=0, degree=1, reorder=None,
                                  comm=comm)
 
 
+def OctahedralSphereMesh(radius, refinement_level=0, degree=1, reorder=None,
+                         comm=COMM_WORLD):
+    """Generate an octahedral approximation to the surface of the
+    sphere.
+
+    :arg radius: The radius of the sphere to approximate.
+         For a radius R the edge length of the underlying
+         octahedron will be.
+
+         .. math::
+
+             a = \\sqrt{2}{R}
+
+    :kwarg refinement_level: optional number of refinements (0 is an
+        octahedron).
+    :kwarg degree: polynomial degree of coordinate space (defaults
+        to 1: flat triangles)
+    :kwarg reorder: (optional), should the mesh be reordered?
+    :kwarg comm: Optional communicator to build the mesh on (defaults to
+        COMM_WORLD).
+    """
+    if refinement_level < 0 or refinement_level % 1:
+            raise RuntimeError("Number of refinements must be a non-negative integer")
+
+    if degree < 1:
+        raise ValueError("Mesh coordinate degree must be at least 1")
+    # vertices of an octahedron of radius 1
+    vertices = np.array([[1.0, 0.0, 0.0],
+                         [0.0, 1.0, 0.0],
+                         [0.0, 0.0, 1.0],
+                         [-1.0, 0.0, 0.0],
+                         [0.0, -1.0, 0.0],
+                         [0.0, 0.0, -1.0]])
+    # faces of the base octahedron
+    faces = np.array([[0, 1, 2],
+                      [0, 1, 5],
+                      [0, 2, 4],
+                      [0, 4, 5],
+                      [1, 2, 3],
+                      [1, 3, 5],
+                      [2, 3, 4],
+                      [3, 4, 5]], dtype=np.int32)
+
+    plex = mesh._from_cell_list(2, faces, vertices, comm)
+    plex.setRefinementUniform(True)
+    for i in range(refinement_level):
+        plex = plex.refine()
```

Comment:
The cubed mesh uses an equiangular formula to try to keep the areas of the
quads close together.

--cjc

On 17 June 2016 at 14:33, Andrew T. T. McRae <notifications@github.com>
wrote:

> In firedrake/utility_meshes.py
> <https://github.com/firedrakeproject/firedrake/pull/813#discussion_r67510359>
> :
>
> > +                         [0.0, -1.0, 0.0],
> > +                         [0.0, 0.0, -1.0]])
> > +    # faces of the base octahedron
> > +    faces = np.array([[0, 1, 2],
> > +                      [0, 1, 5],
> > +                      [0, 2, 4],
> > +                      [0, 4, 5],
> > +                      [1, 2, 3],
> > +                      [1, 3, 5],
> > +                      [2, 3, 4],
> > +                      [3, 4, 5]], dtype=np.int32)
> > +
> > +    plex = mesh._from_cell_list(2, faces, vertices, comm)
> > +    plex.setRefinementUniform(True)
> > +    for i in range(refinement_level):
> > +        plex = plex.refine()
>
> For discussion: should we "refine, push out" alternately, rather than
> "refine n times, then push out"? I've copied what we do for the
> IcosahedralSphereMesh at the moment.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/firedrakeproject/firedrake/pull/813/files/cbd8a9865631d430a1deb039a1413e6dfcc7e072#r67510359>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AF0q8jB5GzXtYwZIgfIJVoEWjdHJgTmYks5qMqIpgaJpZM4I4XDL>
> .
>



-- 
http://www.imperial.ac.uk/people/colin.cotter

www.cambridge.org/9781107663916